File: src/api/accounts.ts
/** Account API client. Maps snake_case payloads to camelCase for the UI. */

export interface Account {
  publicId: string;
  isPublic: boolean;
  createdAt: string;
}

export interface AccountWithToken extends Account {
  accessToken: string;
}

export interface CreatedAccount extends AccountWithToken {
  mnemonic: string;
}

const TOKEN_KEY = 'notmice.access_token';

interface AccountViewPayload {
  public_id: string;
  is_public: boolean;
  created_at: string;
}

interface SessionPayload extends AccountViewPayload {
  access_token: string;
  token_type: string;
}

interface CreatedPayload extends SessionPayload {
  mnemonic: string;
}

function apiUrl(path: string): string {
  const base = import.meta.env.VITE_API_BASE_URL ?? '';
  return `${base}${path}`;
}

function mapAccount(payload: AccountViewPayload): Account {
  return {
    publicId: payload.public_id,
    isPublic: payload.is_public,
    createdAt: payload.created_at,
  };
}

async function readError(response: Response): Promise<string> {
  try {
    const body: unknown = await response.json();
    if (body && typeof body === 'object' && 'detail' in body) {
      const detail = (body as { detail: unknown }).detail;
      if (typeof detail === 'string') {
        return detail;
      }
    }
  } catch {
    // Fall through to status text.
  }
  return `Request failed (${response.status})`;
}

async function requestJson<T>(path: string, init: RequestInit): Promise<T> {
  const response = await fetch(apiUrl(path), init);
  if (response.status === 204) {
    return undefined as T;
  }
  if (!response.ok) {
    throw new Error(await readError(response));
  }
  return (await response.json()) as T;
}

function jsonHeaders(token?: string): HeadersInit {
  const headers: Record<string, string> = { 'Content-Type': 'application/json' };
  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }
  return headers;
}

export function readStoredToken(): string | null {
  try {
    return sessionStorage.getItem(TOKEN_KEY);
  } catch {
    return null;
  }
}

export function storeToken(token: string): void {
  sessionStorage.setItem(TOKEN_KEY, token);
}

export function clearStoredToken(): void {
  sessionStorage.removeItem(TOKEN_KEY);
}

export async function createAccount(): Promise<CreatedAccount> {
  const payload = await requestJson<CreatedPayload>('/api/v1/accounts', {
    method: 'POST',
    headers: jsonHeaders(),
    body: JSON.stringify({}),
  });
  return {
    ...mapAccount(payload),
    accessToken: payload.access_token,
    mnemonic: payload.mnemonic,
  };
}

export async function loginWithMnemonic(mnemonic: string): Promise<AccountWithToken> {
  const payload = await requestJson<SessionPayload>('/api/v1/accounts/login', {
    method: 'POST',
    headers: jsonHeaders(),
    body: JSON.stringify({ mnemonic }),
  });
  return {
    ...mapAccount(payload),
    accessToken: payload.access_token,
  };
}

export async function logoutAccount(token: string): Promise<void> {
  await requestJson<void>('/api/v1/accounts/logout', {
    method: 'POST',
    headers: jsonHeaders(token),
  });
}

export async function fetchCurrentAccount(token: string): Promise<Account> {
  const payload = await requestJson<AccountViewPayload>('/api/v1/accounts/me', {
    method: 'GET',
    headers: jsonHeaders(token),
  });
  return mapAccount(payload);
}

export async function updateShareSettings(
  token: string,
  isPublic: boolean
): Promise<Account> {
  const payload = await requestJson<AccountViewPayload>('/api/v1/accounts/me/share', {
    method: 'PATCH',
    headers: jsonHeaders(token),
    body: JSON.stringify({ is_public: isPublic }),
  });
  return mapAccount(payload);
}

File: app/api/accounts.py
"""Pseudonymous account HTTP surface."""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Body, Depends, HTTPException, status
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer

from app.core.deps import get_account_service
from app.domain.accounts import (
    AccountError,
    AccountNotFoundError,
    InvalidCredentialsError,
    InvalidMnemonicError,
    UnauthenticatedError,
    UserRecord,
)
from app.domain.pii import PIIValidationError, reject_pii
from app.domain.schemas import (
    AccountCreatedResponse,
    AccountCreateRequest,
    AccountLoginRequest,
    AccountSessionResponse,
    AccountView,
    ShareSettingsUpdate,
)
from app.services.accounts import AccountService

router = APIRouter(prefix="/api/v1/accounts", tags=["accounts"])
_bearer = HTTPBearer(auto_error=False)


def _http_for(exc: AccountError) -> HTTPException:
    """Map domain errors to HTTP responses without leaking identifiers."""
    if isinstance(exc, InvalidMnemonicError):
        return HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid recovery phrase",
        )
    if isinstance(exc, InvalidCredentialsError):
        return HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid recovery phrase",
        )
    if isinstance(exc, (UnauthenticatedError, AccountNotFoundError)):
        return HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Not authenticated",
            headers={"WWW-Authenticate": "Bearer"},
        )
    return HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Account request failed")


async def get_current_user(
    credentials: Annotated[HTTPAuthorizationCredentials | None, Depends(_bearer)],
    account_service: Annotated[AccountService, Depends(get_account_service)],
) -> UserRecord:
    """Require a valid bearer token and return the current user."""
    token = credentials.credentials if credentials is not None else None
    try:
        return await account_service.authenticate(token)
    except AccountError as exc:
        raise _http_for(exc) from exc


@router.post(
    "",
    response_model=AccountCreatedResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_account(
    account_service: Annotated[AccountService, Depends(get_account_service)],
    payload: Annotated[AccountCreateRequest | None, Body()] = None,
) -> AccountCreatedResponse:
    """Create a pseudonymous participant. The BIP-39 phrase is returned once."""
    body = payload if payload is not None else AccountCreateRequest()
    try:
        reject_pii(body.model_dump())
    except PIIValidationError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Forbidden field",
        ) from exc
    created = await account_service.create()
    return AccountCreatedResponse(
        public_id=created.user.public_id,
        is_public=created.user.is_public,
        created_at=created.user.created_at,
        access_token=created.access_token,
        mnemonic=created.mnemonic,
    )


@router.post("/login", response_model=AccountSessionResponse)
async def login(
    payload: AccountLoginRequest,
    account_service: Annotated[AccountService, Depends(get_account_service)],
) -> AccountSessionResponse:
    """Sign in with the 12-word recovery phrase."""
    try:
        reject_pii(payload.model_dump())
        session = await account_service.login(payload.mnemonic)
    except PIIValidationError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Forbidden field",
        ) from exc
    except AccountError as exc:
        raise _http_for(exc) from exc
    return AccountSessionResponse(
        public_id=session.user.public_id,
        is_public=session.user.is_public,
        created_at=session.user.created_at,
        access_token=session.access_token,
    )


@router.post("/logout", status_code=status.HTTP_204_NO_CONTENT)
async def logout(
    _current: Annotated[UserRecord, Depends(get_current_user)],
) -> None:
    """Acknowledge logout. Tokens are stateless; the client must discard them."""
    return None


@router.get("/me", response_model=AccountView)
async def read_me(
    current: Annotated[UserRecord, Depends(get_current_user)],
) -> AccountView:
    """Return the current account. Never includes the recovery phrase."""
    return AccountView(
        public_id=current.public_id,
        is_public=current.is_public,
        created_at=current.created_at,
    )


@router.patch("/me/share", response_model=AccountView)
async def update_share(
    payload: ShareSettingsUpdate,
    current: Annotated[UserRecord, Depends(get_current_user)],
    account_service: Annotated[AccountService, Depends(get_account_service)],
) -> AccountView:
    """Set the opt-in public sharing flag in ``share_settings``."""
    try:
        reject_pii(payload.model_dump())
        updated = await account_service.set_public(current.id, payload.is_public)
    except PIIValidationError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Forbidden field",
        ) from exc
    except AccountError as exc:
        raise _http_for(exc) from exc
    return AccountView(
        public_id=updated.public_id,
        is_public=updated.is_public,
        created_at=updated.created_at,
    )
